packages/context/src/binding-decorator.ts

```diff
@@ -0,0 +1,215 @@
+// Copyright IBM Corp. 2018. All Rights Reserved.
+// Node module: @loopback/context
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {ClassDecoratorFactory, MetadataInspector} from '@loopback/metadata';
+import {Binding, BindingScope, BindingTemplate, TagMap} from './binding';
+import {Provider} from './provider';
+import {Constructor} from './value-promise';
+
+const BINDING_METADATA_KEY = 'binding.metadata';
+
+type BindingMetadata = {
+  templates: BindingTemplate[];
+  target?: Constructor<unknown>;
+};
+
+const removeNameAndKey: BindingTemplate = binding => {
+  if (binding.tagMap) {
+    delete binding.tagMap.name;
+    delete binding.tagMap.key;
+  }
+};
+
+class BindDecoratorFactory extends ClassDecoratorFactory<BindingMetadata> {
+  mergeWithInherited(inherited: BindingMetadata, target: Function) {
+    if (inherited) {
+      return {
+        templates: [
+          ...inherited.templates,
+          removeNameAndKey,
+          ...this.spec.templates,
+        ],
+        target: this.spec.target,
+      };
+    } else {
+      this.withTarget(this.spec, target);
+      return this.spec;
+    }
+  }
+
+  withTarget(spec: BindingMetadata, target: Function) {
+    spec.target = target as Constructor<unknown>;
+    return spec;
+  }
+}
+
+export type BindingScopeAndTags = {
+  scope?: BindingScope;
+  tags?: string | TagMap | (string | TagMap)[];
```

Comment:
This type definition looks long:

```ts
string | TagMap | (string | TagMap)[];
```

Can we use a named type please? Please check for other places that may be using this union type, and update them to use the new named type too.